Add openssl_privatekey_info module

felixfontein opened the pull request:
##### SUMMARY
Adds a new module `openssl_privatekey_info` module, similar to `openssl_certificate_info` in #54709.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
openssl_privatekey_info

felixfontein committed:
Add openssl_privatekey_info module.

MarkusTeufelberger commented on lib/ansible/modules/crypto/openssl_privatekey_info.py:
```diff
@@ -0,0 +1,351 @@
+            - Set this to C(no) if you do not want any private information about this key
+              leave the remote machine.
+        type: bool
+        default: yes
```
Please leave this defaulting to `no` and let users opt-in to transmitting unencrypted keys to a remote machine instead.

There's far too much risk of private keys ending up in logs etc. and there's very little you can do with raw private keys anyways (the more widely used thing would probably be the associated public key).

felixfontein replied:
Fixed.

MarkusTeufelberger commented on lib/ansible/modules/crypto/openssl_privatekey_info.py:
```diff
@@ -0,0 +1,351 @@
+            path=dict(type='path', required=True),
+            passphrase=dict(type='str', no_log=True),
+            return_private_key_data=dict(type='bool', default=True),
+            valid_at=dict(type='dict'),
```
huh?

felixfontein replied:
Guess where I copied the module from... :D

felixfontein committed:
Addressing review feedback.

felixfontein committed:
Update docs.

felixfontein committed:
Update tests.

felixfontein committed:
Work around too broad sanity checks.

felixfontein committed:
...

felixfontein committed:
Don't die when None is returned.

felixfontein committed:
Use OpenSSL to extract RSA and DSA key data.

felixfontein committed:
Extend tests.

felixfontein committed:
Make OpenSSL code compatible to OpenSSL < 1.1.

felixfontein committed:
Rewrite tests to use result dicts instead of result lists.

felixfontein committed:
Skip ECC for too old PyOpenSSL.

felixfontein commented:
I had to access OpenSSL directly for getting information on pyOpenSSL private keys (newer versions allow to get hold of a `cryptography` private key, which makes extracting information easier). But I think that now it works well enough, and apparently also for older OpenSSL versions (Ubuntu 16.04 and FreeBSD 11.1 have 1.0.x, OSX has 0.9.8zg).

ready_for_review

thomwiggers commented on lib/ansible/module_utils/crypto.py:
```diff
@@ -94,12 +94,19 @@ def get_fingerprint(path, passphrase=None):
+def load_privatekey(path, passphrase=None, check_passphrase=True, path_content=None, backend='pyopenssl'):
+    """Load the specified OpenSSL private key.
+
+    The content can also be specified via path_content; in that case,
```
`content` rather than `path_content` would make more sense to me as an argument name.

felixfontein replied:
Fixed.

thomwiggers commented on lib/ansible/modules/crypto/openssl_privatekey_info.py:
```diff
@@ -0,0 +1,462 @@
+        description:
+            - Whether to return private key data.
+            - Set this to C(yes) if you do want any private information about this key
+              leave the remote machine.
```
I would rephrase this as "Only set this to yes when you want private information about this key to leave the remote machine", the way it's written now I read "don't" instead of "do want"

felixfontein replied:
Done.

thomwiggers commented on lib/ansible/modules/crypto/openssl_privatekey_info.py:
```diff
@@ -0,0 +1,462 @@
+
+            # Fail if no backend has been found
+            if backend == 'auto':
+                module.fail_json(msg=("Can't detect none of the required Python libraries "
```
Should be "Can't detect any" rather than "none" I think.

felixfontein replied:
Fixed (also in other modules with same/similar formulation).

felixfontein committed:
Reformulate.

felixfontein committed:
Improve return_private_key_data docs.

felixfontein committed:
Rename path_content -> content.

felixfontein committed:
Add sample.

MarkusTeufelberger commented:
Is this vulnerable to the trick outlined in https://blog.hboeck.de/archives/888-How-I-tricked-Symantec-with-a-Fake-Private-Key.html? Probably needs a call to https://pyopenssl.org/en/stable/api/crypto.html#OpenSSL.crypto.PKey.check somewhere.

felixfontein commented:
Good question. I'll look into that later. I'll also check what `cryptography` does. (And I guess `openssl_privatekey` also needs that check.)

felixfontein committed:
Cleanup.

felixfontein committed:
Add key consistency check.

felixfontein commented:
@MarkusTeufelberger I've added that check. I found out that `cryptography` does not offer such checks at all (it also doesn't call the OpenSSL `RSA_check_key()` function), so I implemented that by directly calling the OpenSSL function. That won't work if `cryptography` ever supports a non-OpenSSL backend, but that will also break some other stuff (like for `openssl_certificate_info`). I'll probably try to get `cryptography` extended once all the major `_info` modules are done and I know what else is missing :)

felixfontein committed:
Improve description.

felixfontein committed:
Adjust minimal version.

felixfontein committed:
Fallback code for some pyOpenSSL < 16.0 versions.

felixfontein committed:
Also support Ed25519 and Ed448 keys (or not).

felixfontein committed:
Add more consistency checks.

felixfontein committed:
Verify DSA keys manually.

felixfontein committed:
Improve DSA key validation.

felixfontein committed:
Forgot one condition.

felixfontein committed:
Make validation more robust.

felixfontein commented:
ready_for_review

MarkusTeufelberger commented on lib/ansible/modules/crypto/openssl_privatekey_info.py:
```diff
@@ -0,0 +1,667 @@
+    return True
+
+
+def _is_cryptography_key_consistent(key, key_public_data, key_private_data):
```
Maybe these functions should go into `module_utils/crypto`? Seem useful for CSRs and certificates (maybe even SSH keys?) too.

felixfontein replied:
You mean the ones above this? Makes sense. I've added a commit for this; I've kept `_check_dsa_consistency` and `_is_cryptography_key_consistent`, though, since they are (at the moment) private key specific, and the latter specific to `cryptography`. We can still move them when we actually need them in other modules as well (and do some necessary generalizations then).

MarkusTeufelberger replied:
Yes, the ones that verify properties of private/public keys.

I'm not 100% sure what would happen in the `selfsigned` and `ownca` providers if used with a keypair that violates these specs, but I agreee, that can be fixed later.

felixfontein committed:
Move generic arithmetic code to module_utils/crypto.py.

MarkusTeufelberger commented:
Looks great.  
Let's

shipit

felixfontein commented:
@thomwiggers @MarkusTeufelberger thanks for reviewing!
@mkrizek thanks for merging!
